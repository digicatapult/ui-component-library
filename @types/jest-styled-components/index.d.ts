declare module 'jest-styled-components/serializer'
