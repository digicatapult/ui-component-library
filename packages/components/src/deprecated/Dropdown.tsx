import React from 'react'
import styled from 'styled-components'
import Button, {ButtonProps} from './Button'
import {get} from '../constants'
import Details, {DetailsProps} from '../Details'
import getDirectionStyles from '../DropdownStyles'
import useDetails from '../hooks/useDetails'
import sx, {SxProp} from '../sx'
import {ComponentProps} from '../utils/types'

const StyledDetails = styled(Details)`
  position: relative;
  display: inline-block;
`

export type DropdownProps = DetailsProps

const Dropdown = ({children, className, ...rest}: DropdownProps) => {
  const {getDetailsProps} = useDetails({closeOnOutsideClick: true})
  return (
    <StyledDetails className={className} {...getDetailsProps()} {...rest}>
      {children}
    </StyledDetails>
  )
}

export type DropdownButtonProps = ButtonProps

const DropdownButton = ({children, ...rest}: DropdownButtonProps) => {
  return (
    <Button as="summary" aria-haspopup="true" {...rest}>
      {children}
      <DropdownCaret />
    </Button>
  )
}

const DropdownCaret = styled.div<SxProp>`
  border: 4px solid transparent;
  margin-left: 12px;
  border-top-color: currentcolor;
  border-bottom-width: 0;
  content: '';
  display: inline-block;
  height: 0;
  vertical-align: middle;
  width: 0;
  ${sx};
`

type StyledDropdownMenuProps = {
  direction?: 'ne' | 'e' | 'se' | 's' | 'sw' | 'w'
} & SxProp

const DropdownMenu = styled.ul<StyledDropdownMenuProps>`
  background-clip: padding-box;
  background-color: ${get('colors.canvas.overlay')};
  border: 1px solid ${get('colors.border.default')};
  border-radius: ${get('radii.2')};
  box-shadow: ${get('shadows.shadow.large')};
  left: 0;
  list-style: none;
  margin-top: 2px;
  padding: 5px 0 5px 0 !important; //TODO: fix this override on our markdown styles
  position: absolute;
  top: 100%;
  width: 160px;
  z-index: 100;

  &::before {
    position: absolute;
    display: inline-block;
    content: '';
  }

  &::after {
    position: absolute;
    display: inline-block;
    content: '';
  }

  &::before {
    border: 8px solid transparent;
    border-bottom-color: ${get('colors.canvas.overlay')};
  }

  &::after {
    border: 7px solid transparent;
    border-bottom-color: ${get('colors.canvas.overlay')};
  }

  // stylelint-disable-next-line selector-max-type
  > ul {
    list-style: none;
  }
  ${props => (props.direction ? getDirectionStyles(props.theme, props.direction) : '')};
  ${sx};
`

const DropdownItem = styled.li`
  display: block;
  padding: ${get('space.1')} 10px ${get('space.1')} 15px;
  overflow: hidden;
  color: ${get('colors.fg.default')};
  text-overflow: ellipsis;
  white-space: nowrap;
  a {
    color: ${get('colors.fg.default')};
    text-decoration: none;
    display: block;
    overflow: hidden;
    color: ${get('colors.fg.default')};
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  &:focus,
  a:focus {
    color: ${get('colors.fg.onEmphasis')};
    text-decoration: none;
    background-color: ${get('colors.accent.emphasis')};
  }

  &:hover,
  &:hover a {
    color: ${get('colors.fg.onEmphasis')};
    text-decoration: none;
    background-color: ${get('colors.accent.emphasis')};
    outline: none;
  }
  ${sx};
`

DropdownMenu.defaultProps = {direction: 'sw'}
DropdownMenu.displayName = 'Dropdown.Menu'

DropdownItem.displayName = 'Dropdown.Item'

DropdownButton.defaultProps = Button.defaultProps
DropdownButton.displayName = 'Dropdown.Button'

DropdownCaret.displayName = 'Dropdown.Caret'

Dropdown.defaultProps = Details.defaultProps

export type DropdownCaretProps = ComponentProps<typeof DropdownCaret>
export type DropdownMenuProps = ComponentProps<typeof DropdownMenu>
export type DropdownItemProps = ComponentProps<typeof DropdownItem>
/**
 * @deprecated Use ActionMenu instead. See https://primer.style/react/ActionMenu for more details.
 */
export default Object.assign(Dropdown, {
  Caret: DropdownCaret,
  Menu: DropdownMenu,
  Item: DropdownItem,
  Button: DropdownButton,
})
