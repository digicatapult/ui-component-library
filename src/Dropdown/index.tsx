import React from 'react'
import Select, { StylesConfig } from 'react-select'
import styled from 'styled-components'

const Wrapper = styled('div')`
  ${({ width }: { width?: string }) => `
    width: ${width || '50%'}
  `}
`

const ValuesContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
`

const Value = styled('div')`
  padding: 0.5rem;
  padding-left: 0.75rem;
  margin: 0.5rem;
  font-size: 1rem;
  border-radius: 10px 0px 0px 10px;
  color: ${(props: any) => props.textColor || '#216968'};
  background-color: ${(props) => props.color || 'rgba(247, 173, 46, 0.6)'};
  user-select: none;
`

const X = styled('button')`
  all: unset;
  margin-left: 1.3rem;
  color: black;
  transition: fill 0.15s ease-in-out;
  cursor: pointer;
  &:hover {
    color: #bb392d;
  }
  &:focus {
    color: #c82f21;
  }
`

interface Props {
  options?: Array<any>
  styles?: StylesConfig
  selected?: any
  isMulti: boolean
  update: (val: any[]) => void
}

interface IDropdown {
  (args: Props): React.ReactElement
}

const MultiSelect = (props: any) => {
  const { value, onChange } = props

  const handleRemoveValue = (e: any) => {
    if (!onChange) return
    const { name: buttonName } = e.currentTarget
    const removedValue = value.find((val: any) => val.value === buttonName)
    if (!removedValue) return
    props.update(value)
    onChange(
      value.filter((val: any) => val.value !== buttonName),
      { name, action: 'remove-value', removedValue }
    )
  }

  return (
    <Wrapper>
      <Select
        isMulti
        theme={(theme) => ({
          ...theme,
          borderRadius: 0,
        })}
        {...props}
        closeMenuOnSelect={false}
        controlShouldRenderValue={true}
      />
      <ValuesContainer>
        {' '}
        {value.map((val: any) => (
          <Value {...val} key={val.value}>
            {val.label}
            <X name={val.value} onClick={handleRemoveValue}>
              âœ•
            </X>
          </Value>
        ))}
      </ValuesContainer>
    </Wrapper>
  )
}

const Dropdown: IDropdown = ({
  options,
  isMulti = false,
  styles = {
    multiValue: () => ({
      display: 'none',
    }),
    option: (styles) => ({
      ...styles,
      ':hover': {
        ...styles[':hover'],
        backgroundColor: '#B6EFA0',
        color: '#27847A',
      },
    }),
    menuList: (styles) => ({
      ...styles,
      color: '#fff',
      backgroundColor: '#27847A',
    }),
  },
  ...props
}) => {
  const [value, setValue] = React.useState(props.selected)

  const update = (val: any) => props.update(val)
  const onChange = (val: any) => {
    update(val)
    setValue(val)
  }

  // TODO show selected in drop down as well.
  if (isMulti)
    return (
      <MultiSelect
        {...{
          options,
          styles,
          ...props,
          onChange,
          value,
        }}
      />
    )
  return (
    <Wrapper width={'50%'}>
      <Select
        value={props.selected}
        options={options}
        theme={(theme) => ({
          ...theme,
          borderRadius: 0,
        })}
        onChange={onChange}
        isMulti={isMulti}
        closeMenuOnSelect={true}
        styles={styles}
      />
    </Wrapper>
  )
}

export default Dropdown
