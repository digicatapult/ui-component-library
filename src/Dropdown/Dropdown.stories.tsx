import React from 'react'
import Dropdown from './index'
import { Story } from '@storybook/react'

export default {
  title: 'Components/Dropdown',
  component: Dropdown,
  argTypes: {
    isMulti: {
      control: {
        type: 'boolean',
      },
    },
  },
  args: {
    options: [
      {
        value: '1',
        label: 'option 1',
        color: 'rgba(128, 204, 114, 0.6)',
        textColor: 'white',
      },
      {
        value: '2',
        label: 'option 2',
        color: 'rgba(182, 239, 160, 0.6)',
      },
      {
        value: '3',
        label: 'option 3',
        color: 'rgba(223, 230, 103, 0.6)',
      },
      {
        value: '4',
        label: 'option 4',
        color: 'rgba(223, 230, 103, 0.6)',
      },
    ],
  },
}

export const Default: Story<typeof Dropdown> = (args) => {
  const [selected, setSelected] = React.useState({
    value: '4',
    label: 'option 4',
    color: 'rgba(223, 230, 103, 0.6)',
  })

  return (
    <>
      <h1>SELECTED: {JSON.stringify(selected)}</h1>
      <Dropdown
        selected={selected}
        isMulti={false}
        update={(val: any) => setSelected(val)}
        {...args}
      />
    </>
  )
}

export const Multi: Story<typeof Dropdown> = (args) => {
  const [selected, setSelected] = React.useState([
    {
      value: '4',
      label: 'option 4',
      color: 'rgba(223, 230, 103, 0.6)',
    },
  ])

  return (
    <>
      <h1>SELECTED: {selected.map((el) => <p>{JSON.stringify(el)}</p>)}</h1>
      <Dropdown
        selected={selected}
        isMulti={true}
        update={(val: any) => {
          console.log(val)
          setSelected(val)
        }}
        {...args}
      />
    </>
  )
}
