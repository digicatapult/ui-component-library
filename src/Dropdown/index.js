import Dropdown from './Dropdown.tsx'
export default Dropdown
