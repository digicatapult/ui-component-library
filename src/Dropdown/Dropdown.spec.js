import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Dropdown from './Dropdown.tsx'

const options = [
  {
    value: '1',
    label: 'option 1',
    color: 'rgba(128, 204, 114, 0.6)',
    textColor: 'white',
  },
  {
    value: '2',
    label: 'option 2',
    color: 'rgba(182, 239, 160, 0.6)',
  },
  {
    value: '3',
    label: 'option 3',
    color: 'rgba(223, 230, 103, 0.6)',
  },
  {
    value: '4',
    label: 'option 4',
    color: 'rgba(223, 230, 103, 0.6)',
  },
]

describe('Dropdown', () => {
  test('default', () => {
    const tree = renderer
      .create(
        <Dropdown
          isMulti={false}
          update={(val) => console.log(val)}
          options={options}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('default - selected', () => {
    const tree = renderer
      .create(
        <Dropdown
          selected={{
            value: '4',
            label: 'option 4',
            color: 'rgba(223, 230, 103, 0.6)',
          }}
          isMulti={false}
          update={(val) => console.log(val)}
          options={options}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('default - with placeholder', () => {
    const tree = renderer
      .create(
        <Dropdown
          placeholder={'placeholder test'}
          isMulti={false}
          update={(val) => console.log(val)}
          options={options}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('multi - hii', () => {
    const tree = renderer
      .create(
        <Dropdown
          selected={[
            {
              value: '4',
              label: 'option 4',
              color: 'rgba(223, 230, 103, 0.6)',
              details: 'some details...',
            },
          ]}
          isMulti={true}
          update={(val) => console.log(val)}
          options={options}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
