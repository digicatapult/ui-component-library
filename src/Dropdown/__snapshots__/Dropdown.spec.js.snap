// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar default 1`] = `
.c0 {
  width: 50%;
}

<div
  className="c0"
  width="50%"
>
  <div
    className=" css-b62m3t-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className=" css-1im77uy-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className=" css-1fdsijx-ValueContainer"
      >
        <div
          className=" css-1dimb5e-singleValue"
        >
          option 4
        </div>
        <div
          className=" css-qbdosj-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className=""
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className=" css-1hb7zxy-IndicatorsContainer"
      >
        <span
          className=" css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className=" css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar multi - hii 1`] = `
.c0 {
  width: 50%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  padding: 0.5rem;
  padding-left: 0.75rem;
  margin: 0.5rem;
  font-size: 1rem;
  border-radius: 10px 0px 0px 10px;
  color: #216968;
  background-color: rgba(223,230,103,0.6);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3 {
  all: unset;
  margin-left: 1.3rem;
  color: black;
  -webkit-transition: fill 0.15s ease-in-out;
  transition: fill 0.15s ease-in-out;
  cursor: pointer;
}

.c3:hover {
  color: #bb392d;
}

.c3:focus {
  color: #c82f21;
}

<div
  className="c0"
>
  <div
    className=" css-b62m3t-container"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className=" css-1im77uy-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className=" css-3w2yfm-ValueContainer"
      >
        <div
          className=" css-osof68-MultiValueGeneric"
        >
          <div
            className=" css-1ktq6b3-MultiValueGeneric"
          >
            option 4
          </div>
          <div
            aria-label="Remove option 4"
            className=" css-1mbxqx5-MultiValueRemove"
            onClick={[Function]}
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
            role="button"
          >
            <svg
              aria-hidden="true"
              className="css-tj5bde-Svg"
              focusable="false"
              height={14}
              viewBox="0 0 20 20"
              width={14}
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          className=" css-qbdosj-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className=""
            disabled={false}
            id="react-select-3-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className=" css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          className=" css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          className=" css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          className=" css-1xc3v61-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-tj5bde-Svg"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c1"
  >
     
    <div
      className="c2"
      color="rgba(223, 230, 103, 0.6)"
      label="option 4"
      value="4"
    >
      option 4
      <button
        className="c3"
        name="4"
        onClick={[Function]}
      >
        âœ•
      </button>
    </div>
  </div>
</div>
`;
